src/xserver_zmq.cpp

```diff
@@ -7,217 +7,86 @@
 * The full license is in the file LICENSE, distributed with this software. *
 ****************************************************************************/
 
-#include <chrono>
-#include <iostream>
-
-#include "xeus/xguid.hpp"
-#include "xeus-zmq/xauthentication.hpp"
 #include "xeus-zmq/xserver_zmq.hpp"
-#include "xeus-zmq/xmiddleware.hpp"
-#include "xeus-zmq/xzmq_serializer.hpp"
-#include "xpublisher.hpp"
-#include "xheartbeat.hpp"
-#include "xtrivial_messenger.hpp"
+#include "xserver_zmq_impl.hpp"
+#include "xserver_zmq_default.hpp"
+#include "xserver_control_main.hpp"
+#include "xserver_shell_main.hpp"
 
 namespace xeus
 {
 
-    xserver_zmq::xserver_zmq(zmq::context_t& context,
-                             const xconfiguration& config,
-                             nl::json::error_handler_t eh)
-        : m_shell(context, zmq::socket_type::router)
-        , m_controller(context, zmq::socket_type::router)
-        , m_stdin(context, zmq::socket_type::router)
-        , m_publisher_pub(context, zmq::socket_type::pub)
-        , m_publisher_controller(context, zmq::socket_type::req)
-        , m_heartbeat_controller(context, zmq::socket_type::req)
-        , p_auth(make_xauthentication(config.m_signature_scheme, config.m_key))
-        , p_publisher(new xpublisher(context,
-                                     std::bind(&xserver_zmq::serialize_iopub, this, std::placeholders::_1),
-                                     config.m_transport, config.m_ip, config.m_iopub_port))
-        , p_heartbeat(new xheartbeat(context, config.m_transport, config.m_ip, config.m_hb_port))
-        , m_iopub_thread()
-        , m_hb_thread()
-        , p_messenger(new xtrivial_messenger(this))
-        , m_error_handler(eh)
-        , m_request_stop(false)
+    xserver_zmq::xserver_zmq(std::unique_ptr<xserver_zmq_impl> impl)
+        : p_impl(std::move(impl))
     {
-        init_socket(m_shell, config.m_transport, config.m_ip, config.m_shell_port);
-        init_socket(m_controller, config.m_transport, config.m_ip, config.m_control_port);
-        init_socket(m_stdin, config.m_transport, config.m_ip, config.m_stdin_port);
-        m_publisher_pub.set(zmq::sockopt::linger, get_socket_linger());
-        m_publisher_pub.connect(get_publisher_end_point());
-
-        m_publisher_controller.set(zmq::sockopt::linger, get_socket_linger());
-        m_publisher_controller.connect(get_controller_end_point("publisher"));
-        m_heartbeat_controller.set(zmq::sockopt::linger, get_socket_linger());
-        m_heartbeat_controller.connect(get_controller_end_point("heartbeat"));
     }
 
-    // Has to be in the cpp because incomplete
-    // types are used in unique_ptr in the header
-    xserver_zmq::~xserver_zmq() = default;
```

Comment:
You need to keep this in the cpp, otherwise you get either a missing symbol (even a missing vtable in our case), or a crash if it builds.